### ISSUE TITLE
Contribute to storage for the privateuse1 backend.

### ISSUE THREAD

**Issue by wbigat** [pytorch/pytorch#99326]
### 🚀 The feature, motivation and pitch

Hi, 
```class _StorageBase``` is mainly used for the CUDA backend and CPU backend.  Currently  I want to contribute to storage for the privateuse1 backend.  The content to be modified consists of two parts: 

part a: python API, just like  ```pin_memory```

part b: C++ API, just like ```THPStorage_resize```

I don't know if it's better to add a function branch for privateuse1 backend or extract it into a function with distribution capabilities by dispatch.

### Alternatives

For  func  ```def pin_memory(self, device="cuda"):```  to support privateuse1 backend,  an backend-related allocator is need . The possible modification is as follows(example).  

```
    def pin_memory(self, device="cuda"):
        if device == "cuda":
            import torch.cuda
            allocator = torch.cuda.memory._host_allocator()  # type: ignore[attr-defined]
            return type(self)(self.size(), allocator=allocator).copy_(self)
        elif device == torch._C._get_privateuse1_backend_name():
            allocator = torch.privateuse1.memory._host_allocator()
            return type(self)(self.size(), allocator=allocator).copy_(self)
       else:
            raise error
```
In privateuse1.memory:
```
def _host_allocator():
    return torch._C._privateuse1_HostAllocator()

And the completed implementation is implemented by the specific backend, and an allocator is returned.
```

<br/>
<br/>
On the other hand,  <br/>
I want to wrap the cpu storage to a cpu tensor, and use this parameter the param ```device``` as dispatch key for different backbend, such as ```cuda, privateuse1```. just like the fellowing example.

```
    def pin_memory(self, device="cuda"):
            cpu_tensor = empty([]).set_(self)
            set cpu_tensor  dispatch key by device
            return type(self)(cpu_tensor.pin_memory().storage())
```
By this way, no need more function branch for new backend.

### Additional context

Would you please give me some advice? Thanks a lot.  @smessmer  @bdhirsh @ezyang @zou3519

**Comment by jbschlosser** [comment]
Triage review for label

**Comment by ezyang** [comment]
Wrapping and using the tensor operation to implement storage is something we have been doing and I would support here, if it works. Also cc @kurtamohler

**Comment by wbigat** [comment]
There is another case in ```THPStorage_resize```.

```
static PyObject* THPStorage_resize_(PyObject* self, PyObject* number_arg) {
  HANDLE_TH_ERRORS
  const auto& storage = THPStorage_Unpack(self);
  int64_t newsize = THPUtils_unpackLong(number_arg);
  c10::DeviceType device_type = storage.device_type();
  if (device_type == at::kCPU) {
    at::native::resize_bytes_cpu(storage.unsafeGetStorageImpl(), newsize);
#ifdef USE_CUDA
  } else if (device_type == at::kCUDA) {
    ptrdiff_t size_bytes_i = newsize;
    const auto size_bytes = static_cast<size_t>(size_bytes_i);
    at::native::resize_bytes_cuda(storage.unsafeGetStorageImpl(), size_bytes);
#endif
  } else  if (device_type == at::kPrivateUse1) {
    xxx
  }
```

In this case, perhaps trying to wrap storage as a tensor and call ```func: resize_(Tensor(a!) self, *) ``` instead of registering a back-end specialized "[call](at::native::resize_bytes_privateuse1)"  is a better approach?

@ezyang

**Comment by ezyang** [comment]
I generally encourage you to call dispatched tensor operations, which you can then override.
